NativeAOT-LLVM: merge the change to running tests

yowl opened the pull request:
This PR merges up to the change to how the NativeAOT tests are run.  You will see I struggled a bit with how to get this running from a windows.x64 host, and ended up setting `IlcToolsPath` and `IlcSdkPath` which was the least invasive way I thought of.

WIP for yml changes

akoeplinger committed:
Fix operator precedence consistency in ASCIIUtility.GetIndexOfFirstNonAsciiByte (#64814)

The `&&` operator takes precedence over `||`, make it explicit by wrapping in parenthesis which also makes it consistent with other usages like https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8Utility.Transcoding.cs#L887

Sse2 implies little endian so the check only needs to apply to AdvSimd.

AndyAyersMS committed:
JIT: clear lvMustInit from dependent locals (#64877)

Their initialization is handled by the parent local.

Closes #64808.

jkotas committed:
Delete ReflectionModule::m_pCreatingAssembly and PEAssembly::m_creator (#64865)

SingleAccretion committed:
Propagate exception sets for assignments (#64882)

* Propagate exception sets for block assignments

* Propagate exception sets for simple assignments

* Fix fgValueNumberCastHelper

To not rely on VNs of setup/placeholder args.

* Enable the checker

EgorBo committed:
Improvements for SpanHelpers.IndexOf (#64872)

jakobbotsch committed:
Normalize returns after optimizing returned indirs (#64881)

This ensures we normalize returned indirs of locals, even when we fold
it into an access of a promoted local's field. This may change a signed
indir into access of an unsigned field, which may need a sign-extending
cast to be inserted.

It is not ideal that fgMorphRetInd can 'lose' this information, but
given that it is a specialized optimization it seems the simplest
solution is to just rely on the follow-up normalization.

Fix #61359

steveisok committed:
[mobile] Add extra HttpClientHandler linker substitution (#64818)

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>
Co-authored-by: Steve Pfister <steve.pfister@microsoft.com>

MarcoRossignoli committed:
Bump coverlet version to 3.1.2 (#64875)

uweigand committed:
[mono] Remove OP_FCONV_TO_U opcode (#64618)

* Emit OP_FCONV_TO_U8/U4 instead of OP_FCONV_TO_U.

* Remove OP_FCONV_TO_U implementations across all back ends.

* Remove duplicated mono_fconv_u8/u4 icall wrappers.

* Fixes https://github.com/dotnet/runtime/issues/64570

lord-executor committed:
Explicit test for injecting scoped IServiceProvider (#63225) (#64558)

thaystg committed:
[wasm][debugger] Fixing race condition (#64394)

* Fixing race condition.

* Completely avoid race condition as @lewing suggested.

* keeping old name

* Addressing @radical comments

* Addressing @lewing comments.

* fix unrelated change

* Addressing @lewing comment offline.

akoeplinger committed:
Improve backport.yml to only check start of line and instructions for team visibility (#64915)

Before the bot reacted to the /backport command anywhere in the comment, now it needs to be at the start of the line.

Additionally, if the check for repo collaborator fails print a link to the Microsoft team on GitHub to change visibility (most common failure case).

am11 committed:
Turn off -ffp-contract on Unix (#64847)

am11 committed:
Turn off flaky RapidJSON optimization on x64 (#64851)

MaximLipnin committed:
Skip System.Net.NetworkInformation.Tests.PingTest.SendPingWithIPAddressAndBigSize test on Android (#64897)

Addresses #64798

sbomer committed:
Move one more 20H2 queue to server 2022 (#64919)

* Move one more 20H2 queue to server 2022

Fix one more instance not covered by https://github.com/dotnet/runtime/pull/64827.
This should fix the same failure in runtime-extra-platforms.

* Update IsIcuCompatiblePlatform

slask committed:
62606 Return fully qualified path from FileSystemEventArgs (#63051)

* FullPath property now returns fully qualified path

Gets the fully qualified path after combining the directory and name part
as per the thread discussion. Also added and updated tests for the new change.

Fix #62606

* OldFullPath returns fully qualified path

To be consistent with new behavior of FullPath from FileSystemEventArgs
same change was applied for RenamedEventArgs including updated tests.

Fix #62606

* Marked some test cases as Windows OS specific

In the context of using an actual full path
which can vary on Unix vs Windows

Fix #62606

* Using Path.Join instead of the custom Combine

Adjusted unit tests to match the new logic.

Fix #62606

* Account for corner case when fullPath matches root

The previous buggy implementation of RenamedEventArgs
produced a convenient side-effect for PhysicalFilesWatcher.

Without the trailing slash adde by RenamedEventArgs the root
was longer than the fullPath by a trailing slash.

Fix #62606

* Handle relative paths from CWD in given drive

Fix #62606

* Clearly separated unix and windows test cases

Implemented most of the code review suggestions.

Fix #62606

* Use EnsureTrailingSeparator instead of custom code

Also made unit tests more focused by
removing needless asserts from the tests

Fix #62606

* Keeping the old behavior with trailing separator

Reverted the changes to PhysicalFilesWatcher
and added back the trailing separator behavior
as per code review suggestion

Fix #62606

* Used Path.Combien vs PathInternalEnsureSeparator

As per code review suggestions.

Fix #62606

* Added blank line for readability

Pipeline retrigger. Some pipelines timed out, most probably transient issues.

Fix #62606

bholmes committed:
vtable setup fix for generic default interface methods in mono runtime (#64102)

* vtable setup fix for generic default interface methods in mono runtime

When processing the overrides from interface default methods we should
check if the interface class is a generic type definition first and
inflate with the interface class context.

Test case included.

* Update test to ensure that the correct context is used

tmds committed:
Console.Unix: fix missing terminal configuration on SIGINT/SIGQUIT/SIGCONT. (#64200)

The introduction of the managed API for signal handling (PosixSignal)
inadvertently caused terminal configuration to no longer be performed
when no managed handlers are registered.

This adds back the unconditional registration for SIGINT/SIGQUIT/SIGCONT.

The missing registrations can cause the terminal to stop echoing when an
application terminates on Ctrl-C.

dotnet-maestro[bot] committed:
[main] Update dependencies from 8 repositories (#64806)

* Update dependencies from https://github.com/dotnet/arcade build 20220203.1

Microsoft.DotNet.XUnitConsoleRunner , Microsoft.DotNet.CodeAnalysis , Microsoft.DotNet.Build.Tasks.Workloads , Microsoft.DotNet.Build.Tasks.Templating , Microsoft.DotNet.Build.Tasks.TargetFramework , Microsoft.DotNet.Build.Tasks.Packaging , Microsoft.DotNet.Build.Tasks.Installers , Microsoft.DotNet.Build.Tasks.Feed , Microsoft.DotNet.Build.Tasks.Archives , Microsoft.DotNet.Arcade.Sdk , Microsoft.DotNet.ApiCompat , Microsoft.DotNet.XUnitExtensions , Microsoft.DotNet.GenAPI , Microsoft.DotNet.VersionTools.Tasks , Microsoft.DotNet.GenFacades , Microsoft.DotNet.SharedFramework.Sdk , Microsoft.DotNet.RemoteExecutor , Microsoft.DotNet.PackageTesting , Microsoft.DotNet.Helix.Sdk
 From Version 2.5.1-beta.22080.1 -> To Version 2.5.1-beta.22103.1

* Update dependencies from https://github.com/dotnet/xharness build 20220204.1

Microsoft.DotNet.XHarness.CLI , Microsoft.DotNet.XHarness.TestRunners.Common , Microsoft.DotNet.XHarness.TestRunners.Xunit
 From Version 1.0.0-prerelease.22101.2 -> To Version 1.0.0-prerelease.22104.1

* Update dependencies from https://github.com/dotnet/runtime-assets build 20220203.2

Microsoft.DotNet.CilStrip.Sources , System.ComponentModel.TypeConverter.TestData , System.Drawing.Common.TestData , System.IO.Compression.TestData , System.IO.Packaging.TestData , System.Net.TestData , System.Private.Runtime.UnicodeData , System.Runtime.Numerics.TestData , System.Runtime.TimeZoneData , System.Security.Cryptography.X509Certificates.TestData , System.Text.RegularExpressions.TestData , System.Windows.Extensions.TestData
 From Version 7.0.0-beta.22075.1 -> To Version 7.0.0-beta.22103.2

* Update dependencies from https://github.com/dotnet/emsdk build 20220203.4

Microsoft.NET.Workload.Emscripten.Manifest-7.0.100
 From Version 7.0.0-alpha.2.22102.3 -> To Version 7.0.0-alpha.2.22103.4

* Update dependencies from https://github.com/dotnet/hotreload-utils build 20220203.1

Microsoft.DotNet.HotReload.Utils.Generator.BuildTool
 From Version 1.0.2-alpha.0.22081.2 -> To Version 1.0.2-alpha.0.22103.1

* Update dependencies from https://github.com/dotnet/llvm-project build 20220203.3

runtime.win-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools , runtime.win-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk , runtime.linux-arm64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk , runtime.linux-arm64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools , runtime.linux-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk , runtime.linux-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools , runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Sdk , runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.Mono.LLVM.Tools
 From Version 11.1.0-alpha.1.22081.2 -> To Version 11.1.0-alpha.1.22103.3

* Update dependencies from https://github.com/dotnet/icu build 20220203.1

Microsoft.NETCore.Runtime.ICU.Transport
 From Version 7.0.0-preview.2.22101.1 -> To Version 7.0.0-preview.2.22103.1

* Update dependencies from https://github.com/dotnet/emsdk build 20220204.1

Microsoft.NET.Workload.Emscripten.Manifest-7.0.100
 From Version 7.0.0-alpha.2.22102.3 -> To Version 7.0.0-alpha.2.22104.1

* Update dependencies from https://github.com/dotnet/llvm-project build 20220204.2

runtime.win-x64.Microsoft.NETCore.Runtime.ObjWriter , runtime.win-arm64.Microsoft.NETCore.Runtime.ObjWriter , runtime.osx.10.12-x64.Microsoft.NETCore.Runtime.ObjWriter , runtime.osx.11.0-arm64.Microsoft.NETCore.Runtime.ObjWriter , runtime.linux-x64.Microsoft.NETCore.Runtime.ObjWriter , runtime.linux-musl-x64.Microsoft.NETCore.Runtime.ObjWriter , runtime.linux-musl-arm64.Microsoft.NETCore.Runtime.ObjWriter , runtime.linux-arm64.Microsoft.NETCore.Runtime.ObjWriter
 From Version 1.0.0-alpha.1.22073.1 -> To Version 1.0.0-alpha.1.22104.2

* Update dependencies from https://github.com/dotnet/runtime build 20220203.2

Microsoft.NETCore.ILAsm , Microsoft.NETCore.DotNetHostPolicy , Microsoft.NETCore.DotNetHost , Microsoft.NETCore.App.Runtime.win-x64 , System.Runtime.CompilerServices.Unsafe , runtime.native.System.IO.Ports , Microsoft.NET.Sdk.IL , System.Text.Json
 From Version 7.0.0-preview.2.22080.2 -> To Version 7.0.0-preview.2.22103.2

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>
Co-authored-by: Přemek Vysoký <premek.vysoky@microsoft.com>

jkotas committed:
Delete unnecessary entries from ApiCompatBaseline.netcoreapp.netstandard.txt (#64928)

stephentoub committed:
Streamline bool.TryParse/Format (#64782)

* Streamline bool.TryParse/Format

* Factor out the failure path and add length checks

jozkee committed:
Add S.S.C.Cose with COSE Sign1 support (#64461)

* Add S.S.C.Cose

* Override Equals in CoseHeaderLabel

* Add SystemFormatsCborVersion to eng/Versions.props

* Address PR feedback

* Remove commented Debugger.Launch from test

* Src:
* Support signing with empty header maps
* Add HashAlgorithmName argument to the granular APIs
Tests:
* Make crypto keys ThreadStatic to avoid race conditions
* Exclude CoseMessage and CoseSign1Message from browser
* Define DefaultKey and DefaultHash to simplify tests
* Overall clean-up

* Use explicit type instead of var

* Tests: Add logic to verify expected headers

* Do not mutate the protectected map when alg header was not specified

* src: Address suggestions

* tests:
* Fix ThreadStatic properties
* Switch from SkipOnPlatformAttribute to IgnoreForCI in csproj

* src: Address style suggestions

* Use ProjectRef instead of PackageRef

* RequiresPreviewFeaturesAttribute must be internal in non-netcoreapp

* Address feedback on CoseHeaderLabel

* src: use RequiresPreviewFeatures as assembly level attribute
tests: add test project to exclusion list of local browser runs

maxkoshevoi committed:
[Group 2] Enable nullable annotations for `Microsoft.Extensions.Caching.Abstractions` (#64018)

* Annotate things

* EvictionCallback is nullable

mapogolions committed:
Preserve existing keys in dictionary (#63834)

* Binding to dictionary doesn't preserve existing keys of origin dictionary

* Attempt to bind existing instance if it is not null

* Remove redundant if-else statement

* Add test for binding to dictionary with enum as key type

* Don't use var unless it is obvious from the right hand side

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

* Explicitly ignore result

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

ReubenBond committed:
Increase default HostOptions.ShutdownTimeout to 30 seconds (#63712)

stephentoub committed:
Remove spin lock from SocketAsyncEventArgs on Windows (#64770)

* Remove spin lock from SocketAsyncEventArgs on Windows

The Windows implementation of SocketAsyncEventArgs has a spin lock to help coordinate between the thread initiating the Winsock operation and the eventual overlapped completion callback.  There are some operations we delay (e.g. registering for cancellation) until after the operation has already been initiated and shown to complete asynchronously rather than synchronously, and as these are being set up after the Winsock call to perform the socket operation, it's possible for the overlapped completion to happen before or while we do that additional cleanup.  This condition was expected to be rare, and a spin lock is used to ensure that if the race condition does occur, the callback waits for the state set up to be complete before continuing.

However, it turns out for certain operations it's actually not that rare.  In particular, it appears that accepts when there's already a pending connection end up frequently completing asynchronously but immediately, which causes this race condition to manifest, and we've seen the spin lock spin so many times that it falls back to an actual sleep that causes unexpected delays.

We can instead just maintain a simple gate that describes whether the launching thread or the callback thread own completion.  The launcher sets up all the state and then tries to transition to set the gate.  Similarly, the first thing the callback does is set the gate (to a packed result in case the launcher needs it).  Whoever gets there second is responsible for handling completion.  If the launching thread is the one that gets there second, it essentially turns the asynchronous operation into a synchronous one, from the perspective of the caller, just as if the operation had completed synchronously.

* Avoid an Interlocked.Exchange in the common case

dotnet-maestro[bot] committed:
[main] Update dependencies from dotnet/linker (#64848)

Co-authored-by: dotnet-maestro[bot] <dotnet-maestro[bot]@users.noreply.github.com>

stephentoub committed:
Fix SymbolixRegexRunner handling of "beginning" (#64946)

The Runner's Go implementation is loading the `runtext` but ignoring `runtextbeg`.  If a non-0 beginning were passed in, the matcher could end up evaluating beginning anchors incorrectly, as it always considers beginning == 0.  This fixes that by slicing the input to ensure the matcher is always called with a span representing the full beginning-end length, such that it's beginning==0 assumption is true.

MichalStrehovsky committed:
NativeAOT ARM64 libraries testing (#64373)

Enable ARM64 libraries testing for NativeAOT.

* Copy the crosstargeting build approach from crossgen2. We'll now build two ILC compilers when targeting ARM64 from x64: an arm64-hosted one that is useless on the build machine, and a x64-hosted one, that will work on the build machine. The `ILCompiler.csproj/props/_crossarch.csproj` projects mirror crossgen's approach.
* Limit the number of libraries tests to one. Hopefully it avoids the build races that we'd see until we can update the repo build to .NET 7 Preview 1.
* Create an ARM64 NativeAOT CI leg that sends the libraries tests to Helix to run.

echesakov committed:
Local heap optimizations  on Arm64 (#64481)

# Local heap optimizations on Arm64

1. When not required to zero the allocated space for local heap (for sizes up to 64 bytes) - do not emit zeroing sequence. Instead do stack probing and adjust stack pointer:

```diff
-            stp     xzr, xzr, [sp,#-16]!
-            stp     xzr, xzr, [sp,#-16]!
-            stp     xzr, xzr, [sp,#-16]!
-            stp     xzr, xzr, [sp,#-16]!
+            ldr     wzr, [sp],#-64
```

2. For sizes less than one `PAGE_SIZE` use `ldr wzr, [sp], #-amount` that does probing at `[sp]` and allocates the space at the same time. This saves one instruction for such local heap allocations:

```diff
-            ldr     wzr, [sp]
-            sub     sp, sp, #208
+            ldr     wzr, [sp],#-208
```

Use `ldp tmpReg, xzr, [sp], #-amount` when the offset not encodable by post-index variant of `ldr`:
```diff
-            ldr     wzr, [sp]
-            sub     sp, sp, #512
+            ldp     x0, xzr, [sp],#-512
```

3. Allow non-loop zeroing (i.e. unrolled sequence) for sizes up to 128 bytes (i.e. up to `LCLHEAP_UNROLL_LIMIT`). This frees up two internal integer registers for such cases:

```diff
-            mov     w11, #128
-                                               ;; bbWeight=0.50 PerfScore 0.25
-G_M44913_IG19:        ; gcrefRegs=00F9 {x0 x3 x4 x5 x6 x7}, byrefRegs=0000 {}, byref, isz
             stp     xzr, xzr, [sp,#-16]!
-            subs    x11, x11, #16
-            bne     G_M44913_IG19
+            stp     xzr, xzr, [sp,#-112]!
+            stp     xzr, xzr, [sp,#16]
+            stp     xzr, xzr, [sp,#32]
+            stp     xzr, xzr, [sp,#48]
+            stp     xzr, xzr, [sp,#64]
+            stp     xzr, xzr, [sp,#80]
+            stp     xzr, xzr, [sp,#96]
```

4. Do zeroing in ascending order of the effective address:

```diff
-            mov     w7, #96
-G_M49279_IG13:
             stp     xzr, xzr, [sp,#-16]!
-            subs    x7, x7, #16
-            bne     G_M49279_IG13
+            stp     xzr, xzr, [sp,#-80]!
+            stp     xzr, xzr, [sp,#16]
+            stp     xzr, xzr, [sp,#32]
+            stp     xzr, xzr, [sp,#48]
+            stp     xzr, xzr, [sp,#64]
```

In the example, the zeroing is done at `[initialSp-16], [initialSp-96], [initialSp-80], [initialSp-64], [initialSp-48], [initialSp-32]` addresses. The idea here is to allow a CPU to detect the sequential `memset` to `0` pattern and switch into write streaming mode.

wfurt committed:
Improve SSL platform detection (#64923)

* CI matrix change: add Windows Server 2022

* Add registry check for ssl3 - tls1.2 tests

* Disable TLS1.3 on framework code

* Fix typo

* revert queue change

Co-authored-by: Jan Jahoda <jajahoda@microsoft.com>

akoeplinger committed:
Fix runtime tests on Android (#64950)

In https://github.com/dotnet/runtime/pull/64744 I changed the name of the env variable used to store the test results directoy but I didn't notice that the runtime tests don't use the same runner.

Fixes https://github.com/dotnet/runtime/issues/64920

MichalStrehovsky committed:
Add user-facing docs and samples for NativeAOT (#64894)

As of https://github.com/dotnet/runtimelab/commit/04b092003db5ba207d4fa3f3becb7f01828bf16c.

lateralusX committed:
[Mono] Fix potential crash when handling image load event. (#64908)

PeterSolMS committed:
Fix card marking issue (#64748)

Fix issue where we didn't clear cards at the ends of regions.

The problem is that when we reach the end of a region during card marking, we clear the cards between card and card_of (end). If card_of (end) is equal to card, we end up clearing nothing at all.

A similar problem occurs when cards happen to be set between end and the end of the reserved memory of the region - these cards would never be cleared.

Both issues don't affect correctness, but they do affect performance for regions with very few cross generation pointers.

Note this change affects only regions - for segments, we stay with code that is equivalent to what we had before.

adamsitnik committed:
Improve Directory.Move and DirectoryInfo.MoveTo (#63675)

* fix the tests: call the virtual Move method, which is overridden in DirectoryInfo_MoveTo class and allows for testing DirectoryInfo.MoveTo as well

* unify the Directory.Move and DirectoryInfo.MoveTo logic, move the common part to FileSystem.MoveDirectory

* remove duplicated check

* there is no need to perform the extra checks on Windows

* check for source file existence only once

* on WASM rename does not fail with ENOTDIR for files that end with /
so we need to handle it before rename is called

adamsitnik committed:
Reduce the cost of creating a MemoryMappedFile (#63790)

* don't fetch the file size more than once when creating a MemoryMappedFile

* use SafeFileHandle instead FileStream

* ensure the handle is disposed when fetching the file size fails

uweigand committed:
[mono] Fix is_byreflike_set on big-endian platforms (#64903)

* Use int / gint32 types to avoid C# vs. native type mismatches

* Fixes s390x regression after https://github.com/dotnet/runtime/pull/63985

jkotas committed:
Delete StructLayoutAttribute.Pack adjustment (#64965)

This adjustment was trying to minic adjustment performed by type loader, but the logic has been out-of-sync with the type loader evolution for years.
It does not make sense for custom attribute reading via reflection to perform this adjustment.

Fixes #12480

SingleAccretion committed:
Take into account zero-offset field sequences when propagating locals (#64701)

* Add a test

* Fix the problem

LCL_VAR use VN != VN of its SSA def.

Using one for replacement can run afoul of substituting
into a LCL_VAR that has a zero-offset field sequence on
top and thus duplicate the sequence (as the replacement
def will already have this sequence incorporated into its
VN).

* Fix the same problem in local propagation

Just one diff, missing null check elimination: we propagated a field
sequence for an array address (PtrToArrElem) that later wasn't recognized
as implying non-nullness of another PtrToArrElem to the same array.

AndyAyersMS committed:
JIT: fix morph's tree rethreading/recosting logic (#64954)

Morph might sometimes not rethread/recost changed trees, leading to asserts
downstream as ordered traversals ran across deleted operands.

Fixes #64892

AndyAyersMS committed:
JIT: fix containment safety check in LowerHWIntrinsicGetElement (#64960)

Missing call to IsSafeToContainMem was causing us to mistakenly think
an operand was going to be contained when it wasn't.

Fixes #64918

jakobbotsch committed:
Add JIT support for control-flow guard on x64 and arm64 (#63763)

Add support for generating control-flow guard checks. The support is enabled by a JIT flag or by setting COMPlus_JitForceControlFlowGuard=1.

Co-authored-by: Michal Strehovský <MichalStrehovsky@users.noreply.github.com>

AndyAyersMS committed:
JIT: more stringent interference checks in forward sub (#64933)

Forward sub analysis was missing some cases of call arg interference and so
inviting/allowing morph to do unsafe arg reorderings.

Fix by considering all kinds of local nodes, not just GT_LCL_VAR, as sources
of interference.

Closes #64904.

vcsjones committed:
Test and comment one-shot CFB8 with extra padding

The CFB one shot decryption handled up to block-size padding intentionally. This adds a test and a comment to explain why.

CarlVerret committed:
Including FastFloat in parsing process (#62301)

* Including FastFloat in parsing process

* PR step 1 - adjusting code in regards to received comments.

* DigitsToUInt64 : Parsing batches of 8 digits with SWAR

* Update src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

Co-authored-by: Günther Foidl <gue@korporal.at>

* Update src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

Co-authored-by: Günther Foidl <gue@korporal.at>

* MaxMantissaFastPath fix

* merge problem...

* Revert "merge problem..."

This reverts commit 1b5fd5136a5396c6439ffd47ef46d0c8b3134693.

* removing an extra comparison for fast path.

* According to PR review, all requested changes are in this commit.

* According to new serie of comments on this PR.

* Fixing formatting.

* fixing sentence ending in comments

* Update THIRD-PARTY-NOTICES.TXT

Adding license notice for FastFloat algorithm

* Update src/libraries/System.Private.CoreLib/src/System/Number.NumberToFloatingPointBits.cs

Co-authored-by: Tanner Gooding <tagoo@outlook.com>

* Identation of power of 5 table

* some adjustements according to reviewer's requests.

* Unnecessary assignment of a value to 'exponent'

* removing excedent path for total digits < 7

* removing path for totaldigits <7

* getting rid of unused power of 10 table

* Renaming FastFloat specific values
Specifying infinitePower param as hexa

* renaming some variables and adjusting comments.

* Handle endianness swapping for BigEndian systems.

Co-authored-by: Günther Foidl <gue@korporal.at>
Co-authored-by: Tanner Gooding <tagoo@outlook.com>

elinor-fung committed:
Temporarily disable core dumps when running libraries tests on macOS (#65001)

Commit:
Update Android to r23b image (#64567)

* Update Android to r23b image

* Don't use flag unavailable in r23b

stephentoub committed:
Initial roll out of !! (#64720)

* Bump version of C# compiler

This should be reverted when arcade brings in a new enough compiler.

* Fix a few new warnings from updating compiler

* Enable IDE0190 to employ parameter null checking

* Initial roll out of !!

* Fix more compiler warnings from upgrade, plus some test fixes / reverts

* Fix a few more tests

* Disable profiling test

radical committed:
Add `runtime-wasm` pipeline (#64830)

* Add `runtime-wasm` pipeline

This adds support for a `runtime-wasm` pipeline, which uses
`runtime-extra-platforms.yml`.

The wasm jobs have been moved to `runtime-extra-platforms-wasm.yml`, and
the rest to `runtime-extra-platforms-other.yml`. And the two files are
included conditionally.

Effectively, this still runs all these jobs when
`runtime-extra-platforms` is run (same as before). But it allows running
just the wasm jobs, triggered manually with:

    `/azp run runtime-wasm`

When `runtime-wasm` is manually run:

- LibraryTests will run all the tests on linux+windows
- LibraryTests_EAT will run all the tests on linux
- LibraryTests_AOT will run all the tests on linux+windows
- Wasm.Build.Tests, and Wasm debugger tests will run on linux+windows
- Runtime tests will run on linux

When `runtime-extra-platforms` is manually run:

- LibraryTests will run all the tests, if paths changed
- LibraryTests_EAT will run all the tests on linux, if paths changed
- LibraryTests_AOT will run all the tests on linux+windows, if paths changed
- Wasm.Build.Tests, and Wasm debugger tests will never run
- Runtime tests will never run

On rolling builds, `runtime-extra-platforms` runs by default:

- LibraryTests will run all the tests
- LibraryTests_EAT will run all the tests on linux
- LibraryTests_AOT will run all the tests on linux+windows
- Wasm.Build.Tests, and Wasm debugger tests will never run
- Runtime tests will never run

With feedback from @steveisok, and @safern.

* Flow `isExtraPlatformsBuild` to the new *-wasm, and *-other templates

```
/mnt/vss/_work/1/s/.packages/microsoft.dotnet.helix.sdk/7.0.0-beta.22080.1/tools/Microsoft.DotNet.Helix.Sdk.MultiQueue.targets(18,5): error : You must specify at least one target queue to send a job to helix. Use the HelixTargetQueues property or HelixTargetQueue items. [/mnt/vss/_work/1/s/src/libraries/sendtohelixhelp.proj]
```

Fix suggested by @safern

* Update condition, based on feedback

* Fix typo

* Rename the var `RunSmokeTestsOnly` to *not* the match msbuild property

.. used to control smoke tests. The variables in yml seem to get
converted to environment variables which can affect the build. Like
causing unrelated builds to get `RUNSMOKETESTSONLY=true`, and breaking
the build.

MichalStrehovsky committed:
Change the way we build CoreCLR tests for NativeAOT (#64738)

* Remove NativeAOT testing added in dotnet/runtime#62704 to make room for the different strategy.
* Build NativeAOT as part of test build, not test run

yowl commented:
I removed the multi module tests as it looked like they were removed upstream.

yowl commented:
Cc @dotnet/nativeaot-llvm

jkotas commented:
It looks like the merge commit was squashed. The list of original commits is missing (the previous merge PR https://github.com/dotnet/runtimelab/pull/2136 was done correctly).

Could you please redo the merge without squashing so that we have proper history?

yowl committed:
Merge commit '87777b624b0b8ea295b0fe9cd8505ce262ed75c0' into merge-test-runner2

# Conflicts:
#	src/coreclr/tools/aot/ILCompiler/ILCompiler.csproj
#	src/tests/Common/CLRTest.NativeAot.targets
#	src/tests/Common/Directory.Build.targets
#	src/tests/run.cmd

yowl committed:
resolve conflicts

yowl committed:
fix ups to allow compilation of Ilc and jit

yowl committed:
Include BuildIntegration.proj for wasm as it is required by tests

IlcFrameworkPath -> IlcFrameworkNativePath for native libs
Allow IlcSdkPath and IlcToolsPath to be set for cross running of browser wasm tests - assumes build host is the test runner

yowl committed:
always set IlcToolsPath and IlcSdkPath for anything other than wasm, as it was essentially.

yowl committed:
remove multimodule tests

yowl commented:
Reapplied merge without fast-forward

SingleAccretion commented on docs/workflow/building/coreclr/nativeaot.md:
````diff
@@ -68,8 +68,8 @@ Working on the Jit itself, one possible workflow is taking advantage of the gene
 It is also possible to publish an ordinary console project for Wasm using packages produced by the build: `build nativeaot.packages && build nativeaot.packages -a wasm -os Browser`, assuming all the binaries mentioned above have been built (note that the order is important - the build always produces an architecture-independent package that has a dependency on an architecture-dependent one, and w ...
 ```xml
 <ItemGroup>
-  <PackageReference Include="Microsoft.DotNet.ILCompiler.LLVM" Version="7.0.0-dev" />
-  <PackageReference Include="runtime.win-x64.Microsoft.DotNet.ILCompiler.LLVM" Version="7.0.0-dev" />
+  <PackageReference Include="Microsoft.DotNet.ILCompiler.LLVM" Version="7.0.0-*" />
+  <PackageReference Include="runtime.win-x64.Microsoft.DotNet.ILCompiler.LLVM" Version="7.0.0-*" />
````
Did the suffix change for locally built packages?

yowl replied:
I was confused with the docs and the public packages, reverted.

SingleAccretion commented on docs/workflow/building/coreclr/nativeaot.md:
```diff
@@ -57,7 +57,7 @@ For the compilers:
 - Build LLVM either from the command line (`cmake --build . --target LLVMCore LLVMBitWriter`) or from VS 2022. Currently the Jit depends only on the output of LLVMCore and LLVMBitWriter projects.  For the Release configuration, `cmake --build . --config Release  --target LLVMCore LLVMBitWriter`
 - Set the enviroment variable `LLVM_CMAKE_CONFIG` to locate the LLVM config: `set LLVM_CMAKE_CONFIG=path-to-the-build-directory/lib/cmake/llvm`. This location should contain the file `LLVMConfig.cmake`. `LLVM_CMAKE_CONFIG_DEBUG` and `LLVM_CMAKE_CONFIG_RELEASE` can be used instead of `LLVM_CMAKE_CONFIG` to allow the build to select the config file based on the configuration. If set, these variable ...
-- Build the Jits and the ILC: `build clr.jit+clr.wasmjit+clr.nativeaotruntime+clr.nativeaotlibs+clr.tools -c [Debug|Release]`. Note that `clr.jit` only needs to be built once.  Add the `libs` subsets if you want the packages for publishing, e.g. `build clr.jit+clr.wasmjit+clr.nativeaotruntime+clr.nativeaotlibs+clr.tools+libs -c Debug`.
+- Build the Jits and the ILC: `build clr.alljits+clr.wasmjit+clr.nativeaotruntime+clr.nativeaotlibs+clr.tools -c [Debug|Release]`. Note that `clr.jit` only needs to be built once.  Add the `libs` subsets if you want the packages for publishing, e.g. `build clr.alljits+clr.wasmjit+clr.nativeaotruntime+clr.nativeaotlibs+clr.tools+libs -c Debug`.
```
Why is `clr.alljits` needed now?

yowl replied:
Thanks, whatever problem I thought there was does not exist.  Removed change

SingleAccretion commented on eng/Subsets.props:
```diff
@@ -310,6 +310,11 @@
     <ProjectToBuild Include="$(CoreClrProjectRoot)nativeaot\BuildIntegration\BuildIntegration.proj" Category="clr" />
   </ItemGroup>
 
+  <!-- Running the tests for browser wasm requires BuildIntegration -->
+  <ItemGroup Condition="$(_subset.Contains('+clr.nativeaotlibs+')) and ('$(TargetOS)' == 'Browser' and '$(TargetArchitecture)' == 'wasm')">
+    <ProjectToBuild Include="$(CoreClrProjectRoot)nativeaot\BuildIntegration\BuildIntegration.proj" Category="clr" />
+  </ItemGroup>
+
```
The build integration is built for `-a wasm` when we request `clr.nativeaotruntime` and for the host as part of `clr.tools`. Why is it needed in `clr.nativeaotlibs` as well?

yowl replied:
I guess I added that because we dont ask for `clr.nativeaotruntime` at line 52 (hyperlink not working):
```
 Run `build clr.nativeaotlibs+libs -c [Debug|Release] -a wasm -os Browser`
```

yowl replied:
Let me try without it.

yowl replied:
reverted

yowl replied:
Should we add `clr.nativeaotruntime` to line 52 of the nativeaot.md?

```
- Run `build clr.nativeaotlibs+libs -c [Debug|Release] -a wasm -os Browser`. This will create the architecture-dependent libraries needed for linking and runtime execution, as well as the managed binaries to be used as input to ILC. Add the `clr.nativeaotbuild` subset if you want the packages for publishing, e.g. `build clr.nativeaotlibs+clr.nativeaotbuild+libs -c Debug -a wasm -os Browser`
```

Without it you can't run the tests

SingleAccretion replied:
I think #2146 already added it.

SingleAccretion commented on src/coreclr/jit/llvm.cpp:
```diff
@@ -588,7 +588,11 @@ const HelperFuncInfo& Llvm::getHelperFuncInfo(CorInfoHelpFunc helperFunc)
+
+        // TODO-LLVM: validating indirect calls, does LLVM need or want this?
+        { FUNC(CORINFO_HELP_VALIDATE_INDIRECT_CALL) },
+        { FUNC(CORINFO_HELP_DISPATCH_INDIRECT_CALL) }
```
```suggestion
        { FUNC(CORINFO_HELP_PARTIAL_COMPILATION_PATCHPOINT) },
        { FUNC(CORINFO_HELP_VALIDATE_INDIRECT_CALL) },
        { FUNC(CORINFO_HELP_DISPATCH_INDIRECT_CALL) }
```
These `VALIDATE` helpers are part of the CFG feature which is [Windows-only](https://learn.microsoft.com/en-us/windows/win32/secbp/control-flow-guard).

It is possible (probable) that once/if we start thinking about doing non-WASM LLVM targeting we'll need it, but I believe at that point it'd be as simple as adding a flag for LLVM to generate the required code.

SingleAccretion commented on src/tests/Common/Directory.Build.targets:
```diff
@@ -171,7 +171,6 @@
 
-<<<<<<< HEAD
     <!-- When cross-compiling tests (such as is the case for Wasm), we depends on the host's ILC being built. -->
     <!-- This is a workaround for ILC's project not supporting cross-targeting. Remove once that's implemented. -->
```
```suggestion
    <!-- TODO-LLVM: investigate refactoring in light of ILC project supporting cross-targeting. -->
```
Evidently we now do have cross-compilation for ILC, but it is not totally clear to me we can use it "as-is" (if at all).

SingleAccretion commented on src/tests/build.proj:
```diff
@@ -483,6 +483,10 @@
       <GroupBuildCmd Condition="'$(IlcMultiModule)' == 'true'">$(GroupBuildCmd) "/p:IlcMultiModule=true"</GroupBuildCmd>
       <GroupBuildCmd Condition="'$(BuildNativeAotFrameworkObjects)' == 'true'">$(GroupBuildCmd) "/p:BuildNativeAotFrameworkObjects=true"</GroupBuildCmd>
+
+      <!-- Set Ilc and the SDK patch for cross compilation -->
```
Am I right in understanding this is added here because `src/tests/Directory.Build.targets` doesn't set the properties for WASM?

yowl replied:
I think you can remove the conditions from `src/tests/Directory.Build.targets` and get past the missing `IlcSdkPath`.

You then hit 
```
  E:\GitHub\runtimelab2\.dotnet\sdk\6.0.100\MSBuild.dll -distributedlogger:Microsoft.DotNet.Tools.MSBuild.MSBuildLogger,E:\GitHub\runtimelab2\.dotnet\sdk\6.0.100\dotnet.dll*Microsoft.DotNet.Tools.MSBuild.MSBuildForwardingLogger,E:\GitHub\runtimelab2\.dotnet\sdk\6.0.100\dotnet.dll -maxcpucount /maxcpucount -verbosity:m /bl:E:\GitHub\runtimelab2\artifacts\/log/Debug/InnerManagedTestBuild.3.binlog /nodeReuse:false /p:TargetArchitecture=wasm /p:Configuration=Debug /p:LibrariesConfiguration=debug /p:TargetOS=Browser /p:RuntimeOS= /p:RuntimeFlavor=coreclr /p:RuntimeVariant= /p:CLRTestBuildAllTargets= /p:__TestGroupToBuild=3 /p:__SkipRestorePackages=1 /p:DefaultBuildAllTarget=BuildNativeAot /t:Build E:\GitHub\runtimelab2\src\tests\build.proj
    XUnitWrapperGenerator -> E:\GitHub\runtimelab2\artifacts\tests\coreclr\Browser.wasm.Debug\Common\XUnitWrapperGenerator\XUnitWrapperGenerator\XUnitWrapperGenerator.dll
    Dataflow -> E:\GitHub\runtimelab2\artifacts\tests\coreclr\Browser.wasm.Debug\nativeaot\SmokeTests\Dataflow\Dataflow\Dataflow.dll
    Generating native code
    '"E:\GitHub\runtimelab2\artifacts\bin\coreclr\Browser.wasm.Debug\ilc\\ilc"' is not recognized as an internal or external command,
    operable program or batch file.
E:\GitHub\runtimelab2\artifacts\bin\coreclr\Browser.wasm.Debug\build\Microsoft.NETCore.Native.targets(292,5): error MSB3073: The command ""E:\GitHub\runtimelab2\artifacts\bin\coreclr\Browser.wasm.Debug\ilc\\ilc" @"E:\GitHub\runtimelab2\artifacts\tests\coreclr/obj/Browser.wasm.Debug/Managed/nativeaot\SmokeTests\Dataflow\Dataflow\native\Dataflow.ilc.rsp"" exited with code 9009. [E:\GitHub\runtimelab2\src\tests\nativeaot\SmokeTests\Dataflow\Dataflow.csproj] [E:\GitHub\runtimelab2\src\tests\build.proj]
```
Because:
![image](https://user-images.githubusercontent.com/2720041/210647550-7c73c2f6-9f59-4249-b58d-c23811830fa4.png)

SingleAccretion replied:
The way I imagine this should work is that when we are cross-compiling the tests (which should always be the case for WASM), the ILC path should be set to the appropriate "host" value.

We used to achieve this with the code in `src/tests/Common/Directory.Build.targets` that constructed the core root. I now see [#64738](https://github.com/dotnet/runtime/pull/64738) actually just deleted that code, so it doesn't work anymore.

One particular piece of experience that needs to work is compiling individual tests without going through `src/tests/build.proj`.

SingleAccretion replied:
I see we already have `CoreCLRCrossILCompilerDir` defined `liveBuilds.targets`, so it should be straightforward to adapt code in `src/tests/Directory.Build.targets` to use it I would hope.

yowl replied:
ok let me look into that, thanks.

yowl replied:
Doesn't look like it gets set properly, will trace that
![image](https://user-images.githubusercontent.com/2720041/210652388-6ad5e4ca-2024-4624-aa7c-8dc47cca3b32.png)

SingleAccretion replied:
I think the value is right, just the location somewhat unusual. What I do not recall is part of which build is this cross-component, host or target.

SingleAccretion replied:
Hmm, so I looked again at this cross-arch stuff and it appeared simpler to me to keep doing what we've been doing with the core root construction code, i. e. pick the host compiler out of host's build.

The cross-arch setup looks inconvenient for our scenario where cross-compilation is the mainline workflow.

yowl replied:
I think `CORE_ROOT` is setup in run.py, is that what you had in mind?

SingleAccretion replied:
I basically meant that we should migrate what we did in `src/tests/Common/Directory.Build.targets` with `<!-- When cross-compiling tests (such as is the case for Wasm), we depends on the host's ILC being built. -->` to here (and delete from there as it is now evidently dead code).

yowl replied:
Thanks, I see better now.  `src/tests/Common/Directory.Build.targets` would be ok perhaps, but the build fails before it gets a chance to run.  The target we want from there is `CopyDependencyToCoreRoot` however that ultimately is `dependsOn` for `GenerateLayout`.  That is a `TestBuildSteps` for `TestBuild`, the entry target, but its after `ManagedBuild` which is the failing step.  I tried to set up another dependency to `CopyDependencyToCoreRoot`:
```
  <Target Name="ManagedBuild"
      DependsOnTargets="CopyDependencyToCoreRoot;BuildManagedTestGroups"
```
 but it runs into problems because the msbuild properties are not inherited to `src/tests/Common/Directory.Build.targets`, the `RepoRoot` is not set, and `<Import Project="$(RepoRoot)eng/liveBuilds.targets" />` fails.  It might be possible to make this approach work, but I left it there.

Have moved the copy into a new `Target` in `src\tests\Directory.Build.targets`

SingleAccretion replied:
I don't think there is a need to move the core root construction code here, just the path setting.

I. e. before what happened was:
```
// 1) Core root is constructed with the paths / copy target
// 2) Compiler from core root is used when running the tests
```
Now what should happen is:
```
// 1) Compiler from the path we provide is used during the build
```
I. e. there is no more copying of ILC to the core root directory.

yowl replied:
```
  <Target Name="NativeAotWasmCopyDependencyToCoreRoot" BeforeTargets="IlcCompile" Condition="'$(TestBuildMode)' == 'nativeaot' and '$(TargetArchitecture)' == 'wasm'">

    <!-- Set IlcToolsPath for cross compilation -->
    <PropertyGroup>
      <IlcToolsPath>$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'artifacts', 'bin', 'coreclr', '$(HostOS).$(BuildArchitecture).$(CoreCLRConfiguration)\ilc'))</IlcToolsPath>
    </PropertyGroup>
    <Error Condition="!Exists('$(IlcToolsPath)')"
           Text="'$(IlcToolsPath)' does not exist. Subsets clr.jit, clr.wasmjit and clr.tools for $(HostOS) $(BuildArchitecture) in $(CoreCLRConfiguration) configuration must be built before building this project." />
  </Target>
```
Added this target which is passing in CI now.

yowl replied:
hold on, let me rename it

SingleAccretion replied:
The name's not quite right but we can fix it separately.

Edit: or not :).

yowl replied:
ok, async human communication, beep.  I changed it.  @jkotas I guess if it it passes this time, then it can go, thanks

yowl replied:
Looks like I need to also fix a conflict

yowl committed:
Update src/tests/Common/Directory.Build.targets

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
Update src/coreclr/jit/llvm.cpp

Co-authored-by: SingleAccretion <62474226+SingleAccretion@users.noreply.github.com>

yowl committed:
remove alljits from docs

yowl committed:
Merge branch 'nativeaot-test-running' of https://github.com/yowl/runtimelab into yowl_nativeaot-test-running2

yowl committed:
revert local package  suffix change in docs

yowl committed:
remove buildintegration project

yowl committed:
Remove wasm conditions to pick up paths

yowl committed:
remove setting of path, and copy the HostCoreCLRArtifactsPath as we did before.

yowl committed:
resolve conflict, accepting HEAD for the pool

yowl committed:
Dont copy files, just set path

yowl committed:
Change before target to IlcCompile for CI

yowl committed:
rename target to better fit what it does.

yowl committed:
Merge branch 'feature/NativeAOT-LLVM' into nativeaot-test-running

yowl commented:
Passed. thanks for the help.
